### PR TITLE
Only inject metadata for outputs that cannot be reconstructed by skyframe later

### LINKED ISSUE #16789: Directory artifact not present in sandbox using Bazel 6.0.0rc2
## Description of the bug:

Using Bazel 6.0.0rc1 or rc2 and with `--remote_download_outputs=toplevel` (or `minimal`), `remote_cache` (or `disk_cache`) & sandboxed execution, directory artifacts are not present in runfiles in the sandbox.

### Conditions for issue

- Bazel 6.0.0rc1 or rc2
- directory artifact (created with `ctx.actions.declare_directory`) in runfiles of a test or binary target
- sandboxed execution
- `--remote_cache` (or `--disk_cache` enabled)
- `--remote_download_outputs=toplevel` (or `minimal`)

### Issue

The output directory is not present in the sandbox when the test or binary target is run.

## What's the simplest, easiest way to reproduce this bug? Please provide a minimal example if possible.

Clone https://github.com/gregmagolan/bazel_6_remote_download_outputs_repro

Run `./repro.sh` to run a test that fails if the directory artifact is not present in runfiles. Removing the `--remote_download_outputs` flag -or- running with `--spawn_strategy=local` makes the test pass.

## Which operating system are you running Bazel on?

MacOS (but the issue has also been observed on Linux CI)

## What is the output of `bazel info release`?

release 6.0.0rc2

## If `bazel info release` returns `development version` or `(@non-git)`, tell us how you built Bazel.

_No response_

## What's the output of `git remote get-url origin; git rev-parse master; git rev-parse HEAD` ?

```text
git@github.com:gregmagolan/bazel_6_remote_download_outputs_repro.git
master
fatal: ambiguous argument 'master': unknown revision or path not in the working tree.
Use '--' to separate paths from revisions, like this:
'git <command> [<revision>...] -- [<file>...]'
ac7517ea0ef1dad51847b40899c25079b427fb6e
```

## Have you found anything relevant by searching the web?

No

## Any other information, logs, or outputs that you want to share?

First observed when updating [rules_js](https://github.com/aspect-build/rules_js) to Bazel 6.0.0rc1. rules_js makes heavy use of directory artifacts for the node_modules virtual store

### PR DESCRIPTION
**Pull request by coeuvre**
Currently, they are symlinks.

For other outputs, let skyframe read action file system to construct the metadata.

Before this change, we inject metadata of symlink outputs, tree outputs and file outputs inside `RemoteActionFileSystem#flush()` if these outputs are stored inside the in-memory fs. If the outputs are somehow generated in the local fs, skyframe will read the fs later to construct metadata for them.

However, for tree outputs, skyframe always create an empty directory before action execution in the in-memory fs. So inside the `flush`, we always inject metadata for it. It means local tree outputs are ignored. We could update the code to also read local file system when reading tree files, but then the problem is how to construct metadata from file status which is well done by skyframe.

So instead of injecting metadata by traversal the filesystem inside `flush`, we just let skyframe to do the job.

Fixes #16789.

### CODE REVIEW

File: src/main/java/com/google/devtools/build/lib/remote/RemoteActionFileSystem.java [hunk 1]
```diff
@@ -332,36 +299,53 @@ protected boolean isExecutable(PathFragment path) throws IOException {
...
+    } catch (FileNotFoundException e) {
+      localException = e;
     }
+
+    if (remoteException == null || localException == null) {
+      return;
+    }
+
+    localException.addSuppressed(remoteException);
+    throw localException;
```
Comment (tjgq) [inline src/main/java/com/google/devtools/build/lib/remote/RemoteActionFileSystem.java:333]: Nit: move this either before or after all of the methods that use it.

File: src/main/java/com/google/devtools/build/lib/remote/RemoteActionFileSystem.java [hunk 2]
```diff
@@ -332,36 +299,53 @@ protected boolean isExecutable(PathFragment path) throws IOException {
...
   }
 
   @Override
   protected void chmod(PathFragment path, int mode) throws IOException {
-    RemoteFileArtifactValue m = getRemoteMetadata(path);
-    if (m == null) {
-      super.chmod(path, mode);
-    }
+    doOnRemoteAndLocal(path, p -> p.chmod(mode));
   }
```
Comment (tjgq) [inline src/main/java/com/google/devtools/build/lib/remote/RemoteActionFileSystem.java:349]: Do you mind adding some tests covering the various cases (both exist, none exist, only local, only remote)? This will make future regressions easier to catch since we won't have to debug them at  the Skyframe level :)

I think it's fine to only add tests for `chmod` and trust that the `set*` methods are correct, since they're structurally similar.

File: src/main/java/com/google/devtools/build/lib/remote/RemoteActionFileSystem.java [hunk 3]
```diff
@@ -128,41 +129,7 @@ void flush() throws IOException {
...
-          metadataInjector.injectTree(parent, tree.build());
-        }
-      } else {
-        RemoteFileInfo remoteFile =
-            remoteOutputTree.getRemoteFileInfo(path, /* followSymlinks= */ true);
-        if (remoteFile != null) {
-          metadataInjector.injectFile(output, createRemoteMetadata(remoteFile));
-        }
-      }
+      maybeInjectMetadataForSymlink(path, output);
```
Comment (tjgq) [inline src/main/java/com/google/devtools/build/lib/remote/RemoteActionFileSystem.java:132]: Make this a `void` method since we're no longer consuming the return value.

File: src/main/java/com/google/devtools/build/lib/actions/RemoteFileStatus.java [hunk 4]
```diff
@@ -0,0 +1,8 @@
+package com.google.devtools.build.lib.actions;
+
+import com.google.devtools.build.lib.actions.FileArtifactValue.RemoteFileArtifactValue;
+import com.google.devtools.build.lib.vfs.FileStatusWithDigest;
+
+public interface RemoteFileStatus extends FileStatusWithDigest {
```
Comment (tjgq) [inline src/main/java/com/google/devtools/build/lib/actions/RemoteFileStatus.java:6]: Add a comment explaining when it's appropriate to use this.

**Review by tjgq — APPROVED** [review]

File: src/main/java/com/google/devtools/build/lib/remote/RemoteActionFileSystem.java (hunk 2 above)
Comment (coeuvre) [reply src/main/java/com/google/devtools/build/lib/remote/RemoteActionFileSystem.java:349]: My original plan was to extract the changes to RemoteActionFS to another CL and add tests there.

File: src/main/java/com/google/devtools/build/lib/actions/RemoteFileStatus.java (hunk 4 above)
Comment (coeuvre) [reply src/main/java/com/google/devtools/build/lib/actions/RemoteFileStatus.java:26]: Done.

File: src/main/java/com/google/devtools/build/lib/remote/RemoteActionFileSystem.java (hunk 3 above)
Comment (coeuvre) [reply src/main/java/com/google/devtools/build/lib/remote/RemoteActionFileSystem.java:132]: Done.

File: src/main/java/com/google/devtools/build/lib/remote/RemoteActionFileSystem.java (hunk 2 above)
Comment (coeuvre) [reply src/main/java/com/google/devtools/build/lib/remote/RemoteActionFileSystem.java]: Done with cc712ee5278a6b4686e2ac8fcec3bc54e086fe6e.

**Comment by gregmagolan** [comment]
Confirmed that this change fixes the problem observed on rules_js with Bazel 6.0.0rc1 & rc2. https://github.com/aspect-build/rules_js/pull/659 updates rules_js to `last_green` and CI is green with `--remote_download_toplevel` set.

Thanks for the quick fix @coeuvre !
